### ISSUE TITLE
[BUG] `qml.counts` of a mid-circuit measurement does not return all outcomes even when requested

### ISSUE THREAD

**Issue by astralcai** [PennyLaneAI/pennylane#6700]
### Expected behavior

`qml.counts` with `all_outcomes=True` should return all outcomes.

### Actual behavior

The correct result is returned with deferred measurement.
```python
@qml.qnode(device=dev, mcm_method="deferred")
def circuit():
    m = qml.measure(0)
    return qml.counts(m, all_outcomes=True)
```
```pycon
>>> circuit()
{0: 10, 1: 0}
```
With `"one-shot"`, we get:
```python
@qml.qnode(device=dev, mcm_method="one-shot")
def circuit():
    m = qml.measure(0)
    return qml.counts(m, all_outcomes=True)
```
```pycon
>>> circuit()
{0.0: 10}
```
We see that:
1. We get a float `0.0` instead of the expected integer `0`.
2. We only get one outcome.

### System information

```shell
Name: PennyLane
Version: 0.40.0.dev31
Summary: PennyLane is a cross-platform Python library for quantum computing, quantum machine learning, and quantum chemistry. Train a quantum computer the same way as a neural network.
Home-page: https://github.com/PennyLaneAI/pennylane
Author: 
Author-email: 
License: Apache License 2.0
Location: /Users/REDACTED/pennylane/venv/lib/python3.10/site-packages
Requires: appdirs, autograd, autoray, cachetools, diastatic-malt, networkx, numpy, packaging, pennylane-lightning, requests, rustworkx, scipy, tomlkit, typing_extensions
Required-by: PennyLane-Catalyst, PennyLane_Lightning, PennyLane_Lightning_Kokkos

Platform info:           macOS-15.1.1-arm64-arm-64bit
Python version:          3.10.14
Numpy version:           1.26.4
Scipy version:           1.12.0
Installed devices:
- default.clifford (PennyLane-0.40.0.dev32)
- default.gaussian (PennyLane-0.40.0.dev32)
- default.mixed (PennyLane-0.40.0.dev32)
- default.qubit (PennyLane-0.40.0.dev32)
- default.qutrit (PennyLane-0.40.0.dev32)
- default.qutrit.mixed (PennyLane-0.40.0.dev32)
- default.tensor (PennyLane-0.40.0.dev32)
- null.qubit (PennyLane-0.40.0.dev32)
- reference.qubit (PennyLane-0.40.0.dev32)
- nvidia.custatevec (PennyLane-Catalyst-0.9.0.dev41)
- nvidia.cutensornet (PennyLane-Catalyst-0.9.0.dev41)
- oqc.cloud (PennyLane-Catalyst-0.9.0.dev41)
- softwareq.qpp (PennyLane-Catalyst-0.9.0.dev41)
- lightning.kokkos (PennyLane_Lightning_Kokkos-0.39.0.dev38)
- lightning.qubit (PennyLane_Lightning-0.39.0)
```

### Existing GitHub issues

- [X] I have searched existing GitHub issues to make sure the issue does not already exist.
